### PR TITLE
Add instance name resolution

### PR DESCRIPTION
**Pull request by antho1404**
- Possibility to resolve an instance hash based on a sid/service hash
- use resolution in the application
- introduction of the PromiseStream in order to keep listenResult and listenEvent synchrone and still have the asynchronous resolution

### CODE REVIEW

**Comment by NicolasMahe** [comment]
The name resolution is a good system but it's directly implemented on:
```js
mesg.listenEvent({
  filter: {
    instanceHash: 'emit-event-interval'
  }
})
```

what about add a new parameter in filter that dev should use to use this instance hash resolver? something like:
```js
mesg.listenEvent({
  filter: {
    sid: 'emit-event-interval'
  }
})
```

I also don't see the point to resolve service hash. the system will be much simpler if it only resolve sid when the parameter sid is set (and instanceHash in empty).

File: src/util/resolve.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+// findInstanceFromService returns an instance of a service with the following strategy:
+//   - find a service that has the same hash than the serviceID
+//   - find a service that has the same sid than the serviceID (only if there is no match for the hash)
+//   - get all the instances of this service
+//   - return the instance of the service and throw an error if there is other than one instance running
+const findInstanceFromService = async (api: API, serviceID: string): Promise<Instance> => {
+  const { services } = await api.service.list({})
+  // Find by hash then by sid
+  const service = services.find(x => x.hash === serviceID) || services.find(x => x.sid === serviceID)
```
[the remark this answers is not in the document]
Comment (antho1404) [reply src/util/resolve.ts]: I worked again on this function, there is only sid resolution so that's a lot simpler

**Comment by antho1404** [comment]
I did your last proposition with a resolve function exposed by the application and it only resolves the sid not the service hash anymore. This is a lot easier

File: src/util/resolve.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+const _resolutionTable: Map<string, hash> = new Map()
+
+// returns an instanceHash based on an sid
+// throw an error if the sid doesn't exists or have if it has more or less than one instance running
+export const resolveSID = async (api: API, sid: string): Promise<hash> => {
+  if (_resolutionTable.has(sid)) return _resolutionTable.get(sid)
+
+  const { services } = await api.service.list({})
+  const service = services.find(x => x.sid === sid)
```
Comment (NicolasMahe) [inline src/util/resolve.ts]: same issue as before, they can be multiple service with the same sid. a filter should be used:
```suggestion
  const services = services.filter(x => x.sid === sid)
  if (services.length > 1) throw new Error(`multiple services resolve ${sid}`)
  const service = services.pop()
```

File: src/util/resolve_test.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import * as test from 'tape'
+import * as sinon from 'sinon'
+import { resolveSID } from './resolve'
+import Api from '../api/mock'
+
+const instances = [{ hash: 'instancehash', serviceHash: 'servicehash' }]
+const services = [{ hash: 'servicehash', sid: 'servicesid' }]
```
Comment (NicolasMahe) [inline src/util/resolve_test.ts]: This test have 2 services with same sid:
```suggestion
const services = [{ hash: 'servicehash', sid: 'servicesid' }, { hash: 'servicehash2', sid: 'servicesid' }]
```

File: src/util/resolve.ts (hunk 2 above)
Comment (antho1404) [reply src/util/resolve.ts]: done thx

File: src/util/resolve_test.ts (hunk 3 above)
Comment (antho1404) [reply src/util/resolve_test.ts]: done

File: src/util/resolve.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+// returns an instanceHash based on an sid
+// throw an error if the sid doesn't exists or have if it has more or less than one instance running
+export const resolveSID = async (api: API, sid: string): Promise<hash> => {
+  if (_resolutionTable.has(sid)) return _resolutionTable.get(sid)
+
+  const { services } = await api.service.list({})
+
+  const matching = services.filter(x => x.sid === sid)
+  if (matching.length == 0) throw new Error(`cannot resolve ${sid}`)
```
Comment (NicolasMahe) [inline src/util/resolve.ts]:
```suggestion
  if (matching.length === 0) throw new Error(`cannot resolve ${sid}`)
```

File: src/util/resolve_test.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+});
+
+test('resolve service by sid (multiple)', async function (t) {
+  t.plan(1);
+  const api = Api('')
+  const sid = "multiplesid"
+  sinon.stub(api.instance, 'get').callsFake(() => { throw new Error("not found") })
+  sinon.stub(api.service, 'list').callsFake(() => ({ services: [{ sid }, { sid }] }))
+  try {
+    console.log(await resolveSID(api, sid))
```
Comment (NicolasMahe) [inline src/util/resolve_test.ts]:
```suggestion
    await resolveSID(api, sid)
```

File: src/util/resolve_test.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
```
Comment (NicolasMahe) [inline src/util/resolve_test.ts]:
```suggestion
```

**Pull request merged**